cpp/tests/test_all_gather.cpp

```diff
@@ -0,0 +1,292 @@
+#include <rapidsmpf/buffer/resource.hpp>
+#include <rapidsmpf/communicator/communicator.hpp>
+#include <rapidsmpf/error.hpp>
+#include <rapidsmpf/progress_thread.hpp>
+#include <rapidsmpf/statistics.hpp>
+
+#include "environment.hpp"
+#include "utils.hpp"
+
+using namespace rapidsmpf::all_gather;
+
+extern Environment* GlobalEnvironment;
+
+// Generate a packed data object with the given number of elements and offset.
+// Both metadata and gpu_data contains the same data.
+rapidsmpf::PackedData generate_packed_data(
+    int n_elements,
+    int offset,
+    rmm::cuda_stream_view stream,
+    rapidsmpf::BufferResource& br
+) {
+    auto values = iota_vector<int>(n_elements, offset);
+
+    auto metadata = std::make_unique<std::vector<uint8_t>>(n_elements * sizeof(int));
+    std::memcpy(metadata->data(), values.data(), n_elements * sizeof(int));
+
+    auto data = std::make_unique<rmm::device_buffer>(
+        values.data(), n_elements * sizeof(int), stream, br.device_mr()
+    );
+
+    return {std::move(metadata), br.move(std::move(data), stream)};
+}
+
+// Validate the packed data object by checking the metadata and gpu_data.
+void validate_packed_data(
+    rapidsmpf::PackedData const& packed_data,
+    int n_elements,
+    int offset,
+    rmm::cuda_stream_view stream
+) {
+    auto const& metadata = *packed_data.metadata;
+    EXPECT_EQ(metadata.size(), n_elements * sizeof(int));
+
+    for (int i = 0; i < n_elements; i++) {
+        int val;
+        std::memcpy(&val, metadata.data() + i * sizeof(int), sizeof(int));
+        EXPECT_EQ(offset + i, val);
+    }
+
+    EXPECT_EQ(packed_data.data->size, n_elements * sizeof(int));
+    std::vector<uint8_t> copied_data(n_elements * sizeof(int));
+    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+        copied_data.data(),
+        packed_data.data->data(),
+        n_elements * sizeof(int),
+        cudaMemcpyDefault,
+        stream
+    ));
+
+    RAPIDSMPF_CUDA_TRY(cudaStreamSynchronize(stream));
+    EXPECT_EQ(metadata, copied_data);
+}
+
+class BaseAllGatherTest : public ::testing::Test {
+  protected:
+    void SetUp() override {
+        stream = cudf::get_default_stream();
+        mr = std::unique_ptr<rmm::mr::device_memory_resource>(
+            new rmm::mr::cuda_memory_resource{}
+        );
+        br = std::make_unique<rapidsmpf::BufferResource>(mr.get());
+        comm = GlobalEnvironment->comm_.get();
+
+        all_gather = std::make_unique<AllGather>(
+            GlobalEnvironment->comm_,
+            GlobalEnvironment->progress_thread_,
+            0,
+            stream,
+            br.get()
+        );
+        GlobalEnvironment->barrier();
+    }
+
+    void TearDown() override {
+        GlobalEnvironment->barrier();
+    }
+
+    rmm::cuda_stream_view stream;
+    rapidsmpf::Communicator* comm;
+    std::unique_ptr<rapidsmpf::BufferResource> br;
+    std::unique_ptr<AllGather> all_gather;
+    std::unique_ptr<rmm::mr::device_memory_resource> mr;
```

Comment:
dtor order is the reverse of this. So by the time you come to destruct the allgather and the buffer resource, the memory resource will be gone. This is a problem. You need to make sure that the member order matches the initialisation order requirements

Reply:
Thanks @wence-. I fixed that. But its not only that. I've noticed sometimes, some ranks go into the `TearDown` and start a barrier, while other ranks are still inside the test case. I think let's keep an eye on that. If that comes back again, we'll investigate more. 
I feel like we should differentiate test runs using the OpID. It gives each test case its own communication space.